### PR TITLE
Move offer logic to external contract

### LINKED ISSUE #109: Decrease contributor's voting power/dividends when tokens are offered
Implement custodial contract to vault tokens in the event of internal offering.

### PR DESCRIPTION
**Pull request by vrde**
Close #109

### CODE REVIEW

File: contracts/InternalMarket/InternalMarketBase.sol [hunk 1]
```diff
@@ -0,0 +1,174 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/utils/Context.sol";
+import "hardhat/console.sol";
+
+contract InternalMarketBase is Context {
```
Comment (sirnicolaz) [inline contracts/InternalMarket/InternalMarketBase.sol]: is the Context necessary?

File: contracts/InternalMarket/InternalMarketBase.sol [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        address from,
+        uint256 amount,
+        uint256 createdAt
+    );
+
+    event OfferExpired(uint128 id, address from, uint256 amount);
+
+    event OfferMatched(uint128 id, address from, address to, uint256 amount);
+
+    uint256 public constant WAITING_TIME_EXTERNAL = 7 days;
```
Comment (sirnicolaz) [inline contracts/InternalMarket/InternalMarketBase.sol]: any chance for a name that better indicates what this waiting time is for? a few proposals:
`OFFER_PERIOD`, `OFFER_DURATION`, `TIME_TO_SECONDARY_MARKET`..

File: test/InternalMarket.ts [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    it("should transfer the given amount of token from the erc20", async () => {
+      await internalMarket.makeOffer(1000);
+      expect(token.transferFrom).calledWith(
+        deployer.address,
+        internalMarket.address,
+        1000
+      );
+    });
+
+    /*
```
Comment (sirnicolaz) [inline test/InternalMarket.ts]: remove this

**Comment by sirnicolaz** [comment]
@vrde could you please clarify why there are half the tests that there used to be? the scenarios that have been removed/commented out don't seem to have been reproduced somewhere else

File: test/InternalMarket.ts [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+      });
+
+      it("should allow to withdraw if the amount is equal to the the amount of all expired offers", async () => {
+        await setEVMTimestamp(ts + WEEK + DAY * 3);
+        await internalMarket.connect(alice).withdraw(bob.address, 11 + 25);
+        expect(token.transfer).calledWith(bob.address, 11 + 25);
+      });
+    });
+
+    /*
```
Comment (sirnicolaz) [inline test/InternalMarket.ts]: if this is not necessary, please remove

File: contracts/InternalMarket/InternalMarketBase.sol (hunk 1 above)
Comment (vrde) [reply contracts/InternalMarket/InternalMarketBase.sol]: It's just for consistency with the other contracts, but we can change it.

**Review by sirnicolaz — APPROVED** [review]

**Pull request merged**